src/pool.rs

```diff
@@ -0,0 +1,259 @@
+/// removal will be deferred until all guards are dropped.
+pub struct PoolGuard<'a, T, C>
+where
+    T: Clear + Default,
+    C: cfg::Config,
+{
+    inner: page::slot::Guard<'a, T, C>,
+    shard: &'a Shard<T, C>,
+    key: usize,
+}
+
+impl<T, C> Pool<T, C>
+where
+    T: Clear + Default,
+    C: cfg::Config,
+{
+    /// The number of bits in each index which are used by the pool.
+    ///
+    /// If other data is packed into the `usize` indices returned by
+    /// [`Pool::create`], user code is free to use any bits higher than the
+    /// `USED_BITS`-th bit freely.
+    ///
+    /// This is determined by the [`Config`] type that configures the pool's
+    /// parameters. By default, all bits are used; this can be changed by
+    /// overriding the [`Config::RESERVED_BITS`][res] constant.
+    ///
+    /// [`Config`]: trait.Config.html
+    /// [res]: trait.Config.html#associatedconstant.RESERVED_BITS
+    /// [`Slab::insert`]: struct.Slab.html#method.insert
+    pub const USED_BITS: usize = C::USED_BITS;
+
+    /// Creates a new object in the pool, returning a key that can be used to access it.
+    ///
+    /// If this function returns `None`, then the shard for the current thread is full and no items
+    /// can be added until some are removed, or the maximum number of shards has been reached.
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use sharded_slab::Pool;
+    /// let pool: Pool<String> = Pool::new();
+    /// let mut value = Some(String::from("Hello"));
+    ///
+    /// let key = pool.create(|item| *item = value.take().expect("created twice")).unwrap();
+    /// assert_eq!(pool.get(key).unwrap(), String::from("Hello"));
+    /// ```
+    pub fn create(&self, initilizer: impl FnOnce(&mut T)) -> Option<usize> {
+        let tid = Tid::<C>::current();
+        test_println!("pool: create {:?}", tid);
+        self.shards[tid.as_usize()]
+            .get_initialized_slot(initilizer)
+            .map(|idx| tid.pack(idx))
+    }
+
+    /// Creates a new object in the pool, reusing storage if possible. This method returns a key
+    /// which can be used to access the storage.
+    ///
+    /// If this function returns `None`, then the shard for the current thread is full and no items
+    /// can be added until some are removed, or the maximum number of shards has been reached.
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use sharded_slab::Pool;
+    /// let pool: Pool<String> = Pool::new();
+    /// let value = String::from("Hello");
+    ///
+    /// let key = pool.create_with(value).unwrap();
+    /// assert_eq!(pool.get(key).unwrap(), String::from("Hello"));
+    /// ```
+    pub fn create_with(&self, value: T) -> Option<usize> {
+        let tid = Tid::<C>::current();
+        let mut value = Some(value);
+        test_println!("pool: create_with {:?}", tid);
+        self.shards[tid.as_usize()]
+            .get_initialized_slot(move |item| *item = value)
+            .map(|idx| tid.pack(idx))
+    }
```

Comment:
Maybe a better option would be to rewrite this method to take a reference to a value, and use [`clone_from`](https://doc.rust-lang.org/std/clone/trait.Clone.html#method.clone_from) to insert it. In the worst case (when the type of object in the pool uses the default implementation for `Clone::clone_from`), it would still be clobbering the existing object with a clone of the inserted object, but in many cases (at least for many of `std`'s heap allocated types implementing clone, like `String`, `Vec`, etc) the existing allocation may still be reused.

What do you think about replacing this with something like
```suggestion
    pub fn create_from(&self, value: &T) -> Option<usize> {
        let tid = Tid::<C>::current();
        let mut value = Some(value);
        test_println!("pool: create_with {:?}", tid);
        self.shards[tid.as_usize()]
            .get_initialized_slot(|item| item.clone_from(value)
            .map(|idx| tid.pack(idx))
    }
```